lib/config.js

```diff
@@ -0,0 +1,204 @@
+const path = require('path')
+const yaml = require('js-yaml')
+
+const Rules = require('../rules/rules')
+const RuleInfo = require('./ruleinfo')
+const Fixes = require('../fixes/fixes')
+
+function isAbsoluteURL(url) {
+  try {
+    const u = new URL(url)
+    if (u.host !== '') {
+      return true
+    }
+  } catch (e) {}
+  return false
+}
+
+/**
+ * Load a ruleset config from the specified location.
+ *
+ * @param {string} [configLocation] A URL, file location, or directory containing a repolinter config file
+ * @param {array} [processed] list of config files already processed, used to prevent loops
+ * @returns {Object} The loaded repolinter json config
+ * @throws Will throw an error if unable to parse config or if config is invalid
+ */
+async function loadConfig(configLocation, processed = []) {
+  if (configLocation == null) {
+    throw new Error('must specify config location')
+  }
+
+  let configData = null
+  if (isAbsoluteURL(configLocation)) {
+    const res = await fetch(configLocation)
+    if (!res.ok) {
+      throw new Error(
+        `Failed to fetch config from ${configLocation} with status code ${res.status}`
+      )
+    }
+    configData = await res.text()
+  } else {
+    const st = await fs.promises.stat(configLocation)
+    if (st.isDirectory()) {
+      // try to find a repolinter config file in directory
+      configLocation =
+        findConfig('repolint.json', { cwd: configLocation }) ||
+        findConfig('repolint.yaml', { cwd: configLocation }) ||
+        findConfig('repolint.yml', { cwd: configLocation }) ||
+        findConfig('repolinter.json', { cwd: configLocation }) ||
+        findConfig('repolinter.yaml', { cwd: configLocation }) ||
+        findConfig('repolinter.yml', { cwd: configLocation }) ||
+        path.join(__dirname, '../rulesets/default.json')
+    }
+    configData = await fs.promises.readFile(configLocation, 'utf-8')
+  }
+
+  let ruleset
+  // try parsing as JSON, then YAML
+  try {
+    ruleset = JSON.parse(configData)
+  } catch (je) {
+    try {
+      ruleset = yaml.safeLoad(configData)
+    } catch (ye) {
+      throw new Error(
+        `unable to parse ${configLocation} as either JSON (error: ${je}) or YAML (error: ${ye})`
+      )
+    }
+  }
+
+  // merge extended rulesets
+  if (ruleset.extends) {
+    processed.push(configLocation)
+    if (processed.length > 20) {
+      // safeguard against infinite loops. expose as flag one day if needed
+      throw new Error('exceeded maximum 20 ruleset extensions')
+    }
+
+    let parent
+    if (isAbsoluteURL(ruleset.extends)) {
+      parent = ruleset.extends
+    } else if (isAbsoluteURL(configLocation)) {
+      parent = new URL(ruleset.extends, configLocation)
+    } else {
+      parent = path.resolve(path.dirname(configLocation), ruleset.extends)
+    }
+    if (!processed.includes(parent)) {
+      const parentRuleset = await loadConfig(parent, processed)
+      ruleset = lodash.merge({}, parentRuleset, ruleset)
+    }
+  }
```

Comment:
If I'm understanding this correctly, adding a `targetDirectory` parameter would also remove this logic, since you could resolve the relative paths using it at the top of `loadConfig`.

Reply:
I guess it depends on what you meant by the comment above, which I was a little unclear on.  I don't think it removes the need for this entirely, since they method of retrieving the "current" file controls how relative `extend` values are interpreted, as either relative URL or filepath.  At this point in the logic, configLocation is always a URL of file, never a directory.

Reply:
I think we can resolve this now.